### ISSUE TITLE
Trim whitespace - Default enabled? (Data manipulation enabled by default?)

### ISSUE THREAD

**Issue by JaisEdelmann** [litedb-org/LiteDB#822]
It seem's that by default LiteDb trims whitespace on data saved, this has caused us a bit of headache trying and figure out whats going on. 

I don't think it's a great idea that by default the database is actually manipulating the data you are expecting you are saving.

By default this really should be disabled, if you ask me LiteDb should not have any hidden data manipulation :)

**Comment by mbdavid** [comment]
Hi @JaisEdelmann, yep, it's a very old decision based in reduce document size as much as possible during serialization.

This 3 options has same idea: reduce serialization

    /// <summary>
    /// Indicate that mapper do not serialize null values (default false)
    /// </summary>
    public bool SerializeNullValues { get; set; }

    /// <summary>
    /// Apply .Trim() in strings when serialize (default true)
    /// </summary>
    public bool TrimWhitespace { get; set; }

    /// <summary>
    /// Convert EmptyString to Null (default true)
    /// </summary>
    public bool EmptyStringToNull { get; set; }

I also based this decision in SQL Server (that I use everyday work)  that trim results before insert into a VARCHAR column.

**Comment by oholsen** [comment]
I agree with the above: it is better to optimize when one knows it is safer, rather to enable the expected.

**Comment by nightroman** [comment]
I was about to open a bug... but found this and several closed issues with the same explanation "set `TrimWhitespace` to false".

https://github.com/mbdavid/litedb/issues?q=TrimWhitespace

I wonder if anybody ever asked "Why strings are not trimmed?" :)

> I also based this decision in SQL Server (that I use everyday work) that trim results before insert into a VARCHAR column.

Is this really true? Maybe, I do not know, but I am surprised then. But it does not trim TEXT.
E.g. I have just tried SQLite ("LiteDB in SQL world"), it does not trim NTEXT.
MongoDB (LiteDB inspiration) does not trim strings, right?

I think this design decision is very unfortunate.
It is easy to work around and configure, indeed.
But many **are** trapped first, maybe with some pain.

This design decision is not consistent, too. This rule does not apply to stored
`BsonDocument`s. In my scenarios some data come to the DB as typed (trimmed)
and some **logically same** come as documents (not trimmed). Then they are
queried back with inconsistent spaces.

Why strings in user data have leading and trailing spaces in the first place?
I would say, because they are needed.

**Comment by nightroman** [comment]
Maybe a dedicated poll can help to find out what users want?
(Unless this is the final authors preference which we should respect of course.)

**Comment by Joy-less** [comment]
I agree that this is a strange default for LiteDB, since whitespace is generally considered significant in C# strings. I would like to change this functionality in future LiteDB versions, even if it breaks backwards compatibility.

**Comment by hez2010** [comment]
Hitting this today while building RAG backed with LiteDB. With this default splitted text chunks cannot be restored into the original full text correctly. Without proper unit tests it's even quite hard to notice such an issue exists in LiteDB.
Together with `EmptyStringToNull` being true by default makes my app crash randomly whenever there's a text chunk contains whitespaces only.

Can we change the default of both `TrimWhitespace` and `EmptyStringToNull` to `false` in LiteDB 6.0 so that this won't confuse people anymore?

**Comment by JKamsker** [comment]
I remember debugging this exact thing and ragequitting because of headaches. Feel free to look into it yourself - maybe you find a way to fix this and not breaking everything else 😅  @hez2010

**Comment by JKamsker** [comment]
This is intended: by default `BsonMapper` trims whitespace from strings and stores empty strings as `null` when it serializes a class. Both behaviours can be turned off:

```csharp
var mapper = new BsonMapper
{
    TrimWhitespace = false,
    EmptyStringToNull = false,
};
using var db = new LiteDatabase("Filename=app.db", mapper);
```

(or set the same properties on `BsonMapper.Global`). Changing the defaults would silently change how existing data is written for everyone, so we're keeping them as they are and closing this as not planned. If this doesn't cover your case, comment here and we'll reopen it.
